plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsLocalIndexExporter.java

```diff
@@ -0,0 +1,216 @@
+            .settings(Settings.builder().put("index.hidden", false).build());
+        client.admin().indices().create(createIndexRequest, listener);
+    }
+
+    /**
+     * Drop the local OpenSearch Index created by the exporter
+     *
+     * @param listener the listener to be notified upon completion
+     * @throws IOException if an error occurs
+     */
+    private synchronized void dropLocalIndex(ActionListener<AcknowledgedResponse> listener) throws IOException {
+        DeleteIndexRequest deleteIndexRequest = new DeleteIndexRequest(this.getIdentifier());
+        client.admin().indices().delete(deleteIndexRequest, listener);
+    }
+
+    /**
+     * Get the index mapping of the local index
+     *
+     * @return String to represent the index mapping
+     * @throws IOException if an error occurs
+     */
+    private String getIndexMappings() throws IOException {
+        return new String(Objects.requireNonNull(this.localIndexMapping).readAllBytes(), Charset.defaultCharset());
+    }
+
+    /**
+     * Bulk ingest the data into to the predefined local OpenSearch Index
+     *
+     * @param records the data to export
+     * @throws IOException if an error occurs
+     */
+    private synchronized void bulkRecord(List<T> records) throws IOException {
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .timeout(TimeValue.timeValueSeconds(INDEX_TIMEOUT));
+        for (T record : records) {
+            bulkRequest.add(
+                new IndexRequest(getIdentifier()).source(record.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS))
+            );
+        }
+        client.bulk(bulkRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(BulkResponse response) {
+                if (response.status().equals(RestStatus.CREATED) || response.status().equals(RestStatus.OK)) {
+                    log.debug(String.format(Locale.ROOT, "successfully ingest data for %s! ", getIdentifier()));
+                } else {
+                    log.error(String.format(Locale.ROOT, "error when ingesting data for %s", getIdentifier()));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                log.error(String.format(Locale.ROOT, "failed to ingest data for %s, %s", getIdentifier(), e));
+            }
+        });
+    }
+
+    /**
+     * Index one document to the predefined local OpenSearch Index
+     *
+     * @param record the document to export
+     * @throws IOException if an error occurs
+     */
+    private synchronized void indexRecord(T record) throws IOException {
+        IndexRequest indexRequest = new IndexRequest(getIdentifier()).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .source(record.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS))
+            .timeout(TimeValue.timeValueSeconds(INDEX_TIMEOUT));
```

Comment:
Refreshing on each index request might be expensive? Also, why do we need this method to be synchronized? That will significantly bottleneck the concurrent requests

Reply:
Let me remove this index record function and use bulk request only.